File: csharp/2015/day08a/Program.cs
using System;
using System.IO;

class Program
{
    static void Usage()
    {
        Console.Error.WriteLine("usage: dotnet run <input file>");
        Environment.Exit(1);
    }

    static int Process(string filename)
    {
        int result = 0;

        try
        {
            using ( StreamReader reader = new StreamReader(filename) )
            {
                string? line;
                while ( (line = reader.ReadLine()) != null )
                {
                    int codeLen = line.Length;
                    int memLen = 0;
                    int i = 1;
                    while ( i < line.Length - 1 )
                    {
                        switch ( line[i] )
                        {
                        case '\\':
                            switch ( line[i + 1] )
                            {
                            case '\\':
                            case '\"':
                                i += 2;
                                break;
                            case 'x':
                                i += 4;
                                break;
                            default:
                                i += 1;
                                break;
                            }
                            break;
                        default:
                            i += 1;
                            break;
                        }
                        memLen += 1;
                    }
                    result += codeLen - memLen;
                }
            }
        }
        catch ( IOException e )
        {
            Console.Error.WriteLine("Error reading file: " + e.Message);
            Environment.Exit(1);
        }

        return result;
    }

    static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Usage();
        }

        string filename = args[0];
        int result = Process(filename);
        Console.WriteLine("result = " + result);
    }
}

File: fsharp/2015/day08a/Program.fs
open System
open System.IO

let usage() =
    Console.Error.WriteLine("usage: dotnet run <input file>")
    Environment.Exit(1)

let rec scanChars(chars, i) =
    match chars with
    | [] -> i
    | [_] -> i + 1
    | ch1 :: ch2 :: rest ->
        (match (ch1, ch2) with
         | ('\\', '\\') -> scanChars(rest, i + 1)
         | ('\\', '\"') -> scanChars(rest, i + 1)
         | ('\\', 'x')  -> scanChars(List.tail (List.tail rest), i + 1)
         | _ -> scanChars(ch2 :: rest, i + 1)
        )

let processFile(filename: string) =
    try
        let content = File.ReadAllText(filename)
        content.Split([|'\n'|], StringSplitOptions.RemoveEmptyEntries)
        |> Seq.fold
           (fun acc line ->
               let codeLen = line.Length
               let quoted = line.Substring(1, line.Length - 2)
               let memLen = scanChars(Seq.toList quoted, 0)
               acc + (codeLen - memLen)
           )
           0
    with
    | :? IOException as e ->
        Console.Error.WriteLine("Error reading file: " + e.Message)
        Environment.Exit(1)
        0

[<EntryPoint>]
let main args =
    if args.Length < 1 then
        usage()

    let filename = args.[0]
    let result = processFile filename
    Console.WriteLine("result = " + result.ToString());
    0 
